### PR TITLE
8325949: Create an internal utility method for creating VarHandle instances

### PR DESCRIPTION
**Pull request by minborg**
This PR suggests introducing an internal class in `java.base` to simplify the use of some `MethodHandles.Lookup` operations.

While the utility of the methods might appear to be limited in classes with many static `VarHandle`/`MethodHandle` variables, it should be noted that the class files become smaller and simpler. Here are some examples:

| Class File                                      | Base [Bytes] | Patch [Byte] |
| --------------------------------| ------------- | ------------ |
| FutureTask.class                          |            10,255 |          10,123 |
| AtomicBoolean.class                   |             5,364 |            5,134 |
|AtomicMarkableReference.class |             3,890 |           3,660 |

image

The new `MethodHandlesInternal.class` file is of size 1,952 bytes.

In total for `java.base` we have:

| Build map "jdk"  | Size [Bytes] |
| ---------------| ------------- |
| Base                 |     5,906,457 |
| Patch                |     5,905,487 |
| Delta                 |                940|

For 60 billion instances, this represents > 50 TB.

Tried and passed tier1-3

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325949](https://bugs.openjdk.org/browse/JDK-8325949): Create an internal utility method for creating VarHandle instances (**Enhancement** - P5)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20972/head:pull/20972` \
`$ git checkout pull/20972`

Update a local copy of the PR: \
`$ git checkout pull/20972` \
`$ git pull https://git.openjdk.org/jdk.git pull/20972/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20972`

View PR using the GUI difftool: \
`$ git pr show -t 20972`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20972.diff">https://git.openjdk.org/jdk/pull/20972.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20972#issuecomment-2346961509)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/reflect/MethodHandlesInternal.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * static MethodHandle BAR_HANDLE =
+ *         MethodHandlesInternal.findVirtualOrThrow(MethodHandles.lookup(),
+ *                 Foo.class,"bar",MethodType.methodType(int.class));
+ * }
+ */
+public final class MethodHandlesInternal {
+
+    private MethodHandlesInternal() {}
+
+    public static VarHandle findVarHandleOrThrow(MethodHandles.Lookup lookup, Class<?> recv, String name, Class<?> type) {
```
Comment (RogerRiggs) [inline src/java.base/share/classes/jdk/internal/reflect/MethodHandlesInternal.java:47]: The method naming might benefit from being clear that failure is fatal; never call it with arguments that might fail.
Perhaps even throw java.lang.InternalError. ExceptionInInitializerError might be misconstrued as a some kind of race or being called too early.

**Comment by liach** [comment]
I think we can elide the `Lookup` argument to perform a always-full-permission lookup operation to simplify call sequence; or, to elide the `recv` receiver argument as it is almost always `lookup.lookupClass()` (except for a few lazy holders)

File: src/java.base/share/classes/jdk/internal/reflect/MethodHandlesInternal.java
Comment (liach) [inline src/java.base/share/classes/jdk/internal/reflect/MethodHandlesInternal.java:1]: I would argue this is closer to invoke than reflect, so can consider moving to `sun.invoke` or even create a new `jdk.internal.invoke` package.

**Comment by minborg** [comment]
> I think we can elide the `Lookup` argument to perform a always-full-permission lookup operation to simplify call sequence; or, to elide the `recv` receiver argument as it is almost always `lookup.lookupClass()` (except for a few lazy holders)

If we could access `MethodHandles.Lookup.IMPL_LOOKUP `, we could use this as a lookup and elide the lookup arg. Any ideas to do this in a "polite" way?

**Comment by AlanBateman** [comment]
> If we could access `MethodHandles.Lookup.IMPL_LOOKUP `, we could use this as a lookup and elide the lookup arg. Any ideas to do this in a "polite" way?

Lookup.IMPL_LOOKUP should be guarded closely, I don't think we should remove the need to provide a Lookup with the appropriate access.

File: src/java.base/share/classes/jdk/internal/reflect/MethodHandlesInternal.java
Comment (ExE-Boss) [reply src/java.base/share/classes/jdk/internal/reflect/MethodHandlesInternal.java:1]: I’m in favour of introducing `jdk.internal.invoke`.

**Comment by kaspernielsen** [comment]
This would actually be super useful in a public API. I use the same "trick" in some of my projects. Not because I care too much about the classfile size. But it is just a lot easier on the eye.

**Comment by minborg** [comment]
> This would actually be super useful in a public API. I use the same "trick" in some of my projects. Not because I care too much about the classfile size. But it is just a lot easier on the eye.

While I understand it would be convenient, there is a reason we have checked exceptions in the public APIs. Users can create similar utility methods in their projects if they want to "hide" the exceptions as you say. If you wish to add public methods anyhow, I think we should consider this under a separate issue.

**Comment by liach** [comment]
I think the original poster means to find a VH in the lookup class.

**Comment by kaspernielsen** [comment]
> I think the original poster means to find a VH in the lookup class.

I actually meant what Per is hinting at. Defining static MethodHandle/VarHandle in a class at class initialization time, without the ceremony of handling exceptions and throwing ExceptionInInitializerError. 

I could be mistaken but I think this is probably how most MethodHandle/VarHandles are defined.

**Comment by AlanBateman** [comment]
>  If you wish to add public methods anyhow, I think we should consider this under a separate issue.

I think it's a request for 10+ find methods to sit along the existing methods so it might be confusing too.

I think your proposal looks generally okay, not sure about naming it "MethodHandlesInternal" with "internal" in the package name as it's more like a class of utility methods.

**Comment by minborg** [comment]
It is possible to create a project-specific utility method that can unwrap several methods like this:

```
    @FunctionalInterface
    public interface LookupFunction<R> {
        R apply(MethodHandles.Lookup lookup,
                Class<?> refc,
                String name,
                MethodType type) throws NoSuchMethodException, IllegalAccessException;
    }

    public static <R> R unwrap(LookupFunction<R> lookupFunction,
                               MethodHandles.Lookup lookup,
                               Class<?> refc,
                               String name,
                               MethodType type) {
        try {
            return lookupFunction.apply(lookup, refc, name, type);
        } catch (ReflectiveOperationException e) {
            throw new InternalError(e);
        }
    }

    static final MethodHandle MH_SCALE = MethodHandlesUtil.unwrap(MethodHandles.Lookup::findVirtual,
            MethodHandles.lookup(), MemoryLayout.class, "scale", MethodType.methodType(long.class, long.class, long.class));


```

**Comment by eirbjo** [comment]
> I think your proposal looks generally okay, not sure about naming it "MethodHandlesInternal" with "internal" in the package name as it's more like a class of utility methods.

Perhaps `MethodHandlesSupport`, like `jdk.internal.util.ArraysSupport`?

**Comment by liach** [comment]
I think a better approach to this problem might be converting these static final MethodHandle/VarHandle to class-file constants (Constant_MethodHandle for MH, Constant_Dynamic for VH), see JEP 303.

**Review by RogerRiggs — COMMENTED** [review]
Overall looks good. Having a concise source construct representing these factories will enable jlink or similar pre-processing steps to optimize.

The new class will be lonesome in its own package.
I'd shorten the classname to "MHUtil" making the source lines more readable.

**Comment by dmlloyd** [comment]
In my projects I usually just use the `ConstantBootstraps` method directly, which avoids the try/catch (but does not avoid needing a lookup). In my mind I imagine eventually writing a tool which post-processes my classes and replaces all `private static final Xxx xxx = ConstantBootstraps.xxx()` with dynamic constants (similarly, I guess, to what @liach suggested). But maybe just using the existing `ConstantBootstraps` methods is a good stepping stone towards a JEP-303 kind of solution.

**Comment by minborg** [comment]
I've experimented a bit with using the `ConstantBoostraps::fieldVarHandle` method and here is what it looks like at a call site (example from `CompletableFuture`):

```
RESULT = ConstantBootstraps.fieldVarHandle(l, "result", VarHandle.class, CompletableFuture.class, Object.class);
```

Compared to:

```
RESULT = MhUtil.findVarHandle(l,  "result", Object.class);
```

One advantage is that we can remove two methods from the `MhUtil` class but the drawback is we need to add a lot of more parameters at the call sites. Eventually, both methods will work the same except from the exception handling. So, I think it is better to use the proposed `MhUtil` rather than `ConstantBootstraps`.

**Comment by RogerRiggs** [comment]
The current (MhUtil) source is easier to read. 
So unless ConstantBootstraps could have an equivalent (simple) API that evaluates to a DynamicConstant, keep what you have.

**Review by RogerRiggs — APPROVED** [review]
Looks good.

**Pull request closed without merging**